source/common/upstream/load_balancer_impl.h

```diff
@@ -216,64 +216,122 @@ class ZoneAwareLoadBalancerBase : public LoadBalancerBase {
- * Technique is based on http://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf
+ * and compares number of active requests. Technique is based on
+ * http://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf and is known as P2C (power of
+ * two choices).
  *
- * When any of the hosts have non 1 weight, apply random weighted balancing.
- * Randomly pickup the host and send 'weight' number of requests to it.
- * This technique is acceptable for load testing but
- * will not work well in situations where requests take a long time.
- * In that case a different algorithm using a full scan will be required.
+ * When any hosts have a weight that is not 1, an RR EDF schedule is used. Host weight is scaled
+ * by the number of active requests at pick/insert time. Thus, hosts will never fully drain as
+ * they would in normal P2C, though they will get picked less and less often. In the future, we
+ * can consider two alternate algorithms:
+ * 1) Expand out all hosts by weight (using more memory) and do standard P2C.
+ * 2) Use a weighted Maglev table, and perform P2C on two random hosts selected from the table.
+ *    The benefit of the Maglev table is at the expense of resolution, memory usage is capped.
+ *    Additionally, the Maglev table can be shared amongst all threads.
  */
-class LeastRequestLoadBalancer : public LoadBalancer, ZoneAwareLoadBalancerBase {
+class LeastRequestLoadBalancer : public EdfLoadBalancerBase {
 public:
   LeastRequestLoadBalancer(const PrioritySet& priority_set, const PrioritySet* local_priority_set,
                            ClusterStats& stats, Runtime::Loader& runtime,
                            Runtime::RandomGenerator& random,
-                           const envoy::api::v2::Cluster::CommonLbConfig& common_config);
-
-  // Upstream::LoadBalancer
-  HostConstSharedPtr chooseHost(LoadBalancerContext* context) override;
+                           const envoy::api::v2::Cluster::CommonLbConfig& common_config)
+      : EdfLoadBalancerBase(priority_set, local_priority_set, stats, runtime, random,
+                            common_config) {
+    initialize();
+  }
 
 private:
-  HostSharedPtr last_host_;
-  uint32_t hits_left_{};
+  void refreshHostSource(const HostsSource&) override {}
+  double hostWeight(const Host& host) override {
+    // Here we scale host weight by the number of active requests at the time we do the pick. We
+    // always add 1 to avoid division by 0. Note that if all weights are 1, the EDF schedule is
+    // unlikely to yield the same result as P2C given the lack of randomness as well as the fact
+    // that hosts are always picked, regardless of their current request load at the time of pick.
+    // It might be posible to do better by picking two hosts off of the schedule, and selecting
+    // the one with fewer active requests at the time of selection.
+    return static_cast<double>(host.weight()) / (host.stats().rq_active_.value() + 1);
```

Comment:
Let's say we have host X and host Y. host X has weight 128 and host Y has weight 1. Host X has 128 active requests, host Y has 1. The effective weights for X and Y are 128/129 and 1/2. The ratio between active requests are more important than relative weight. I'm not sure how desirable this is/isn't, just pointing out that this  seems not so intuitive.

Reply:
I see your point, but I'm not sure the best way of modifying. Can you think of a different algorithm that would  be better?

Reply:
Per offline convo I'm going to email the researcher we talked to about this and report back.